### PR TITLE
Put all output binaries into $repo/bin

### PR DESCRIPTION
**Pull request by duglin**
Signed-off-by: Doug Davis <dug@us.ibm.com>

### CODE REVIEW

**Comment by MHBauer** [comment]
I like the name `bin`.

**Comment by duglin** [comment]
actually, never mind about the test request - CI will do it

File: Makefile [hunk 1]
```diff
@@ -32,12 +32,14 @@ SUB := $(addsuffix .sub, $(ALL))
...
 build-linux: build-linux.sub
 build-darwin: build-darwin.sub
-clean: clean.sub
 docker: docker.sub
 lint: lint.sub
 push: push.sub
 test: test.sub
 
+clean: clean.sub
+	rm -rf bin
```
Comment (martinmaly) [inline Makefile]: "$(ROOT)/bin"

I'd probably use that over just 'bin' ... same thing however, so up to you.

File: hack/Common.mk [hunk 2]
```diff
@@ -21,7 +21,8 @@ ifneq (,$(BIN))
 
 build:
 	$(ECHO) echo "Building $(PKG)"
-	$(ECHO) $(GO) build -o "$(BINDIR)/$(BIN)" -ldflags "-X github.com/kubernetes-incubator/service-catalog/pkg.VERSION=$(VERSION)" "$(PKG)"
+	$(ECHO) mkdir -p "$(ROOT)/bin"
+	$(ECHO) $(GO) build -o "$(ROOT)/bin/$(BIN)" -ldflags "-X github.com/kubernetes-incubator/service-catalog/pkg.VERSION=$(VERSION)" "$(PKG)"
```
Comment (martinmaly) [inline hack/Common.mk]: something worth trying maybe to redefine BINDIR in Makefile.mk to `$(ROOT)/bin`

**Comment by martinmaly** [comment]
Seems to work just fine. Wonderful!

**Comment by MHBauer** [comment]
Can I override BINDIR somehow to redirect output anywhere I want?

**Comment by martinmaly** [comment]
I am not seeing a use case for it but it (can you please elaborate)?

the technical solution is possible, change := to ?=  here:
https://github.com/kubernetes-incubator/service-catalog/pull/147/files#diff-42e6e2455c327e591ffa42c2dd88e61eR33

**Comment by MHBauer** [comment]
More out of curiosity than requirement. I've seen the variable `BINDIR` used in a lot of places, and I have at times had it as a shell variable that points to something on my `$PATH`. I think I'm probably confusing shell vars and make vars and how they interact.

**Comment by MHBauer** [comment]
Is this the right result? Should `api` also be executable? 
```
 ls -l bin
total 131688
-rw-r--r--  1 mhb  staff    198154 Dec  9 11:12 api
-rwxr-xr-x  1 mhb  staff  47178740 Dec  9 11:12 controller
-rwxr-xr-x  1 mhb  staff   7223468 Dec  9 11:11 k8s-broker
-rwxr-xr-x  1 mhb  staff   6397004 Dec  9 11:11 registry
-rwxr-xr-x  1 mhb  staff   6415500 Dec  9 11:11 user-broker
```

**Comment by martinmaly** [comment]
The pkg/apis/servicecatalog doesn't actually produce any executable binary (doesn't contain `main`)

It would be more correct to remove this line:
https://github.com/kubernetes-incubator/service-catalog/blob/master/pkg/apis/servicecatalog/Makefile#L15

**Comment by martinmaly** [comment]
Re: variables and interaction: env variables become make variables but (generally), make variables of the same name take precedence. https://www.gnu.org/software/make/manual/make.html#Environment

There are different kinds of assignments, ?= will assign if variable not defined which makes it good of defaults and things such as:

```
make COVERAGE=custom-coverage.html coverage
# or
COVERAGE=custom-coverage.html make coverage
```
https://github.com/kubernetes-incubator/service-catalog/blob/master/hack/Makefile.mk#L35

Make variables are not made automatically visible to recipes unless they are explicitly exported.

**Comment by MHBauer** [comment]
@duglin do you want to handle the additional apiserver output?

**Comment by duglin** [comment]
@MHBauer done

**Comment by MHBauer** [comment]
Can you fix https://github.com/kubernetes-incubator/service-catalog/pull/147#issuecomment-266099327 too?

**Comment by duglin** [comment]
@MHBauer done (again)

**Comment by arschles** [comment]
LGTM

**Comment by MHBauer** [comment]
gonna need to be rebased on the merged #170 first.

**Comment by duglin** [comment]
hmmm github's not complaining

**Comment by MHBauer** [comment]
and the same modification to the makefile added in #158.

**Comment by MHBauer** [comment]
Not rebase because conflicts, rebase because other makefile changes interact and are better done here rather than followups.

**Comment by duglin** [comment]
I think github will do that automagically

**Comment by MHBauer** [comment]
LGTM

**Pull request merged**